src/software/coroutine_main.cpp

```diff
@@ -0,0 +1,452 @@
+    nmo.execute();
+
+    *data = 5;
+
+    fmo.execute();
+    dmo.execute();
+    nmo.execute();
+
+    // Output
+    // 99
+    // 99
+    // 99
+    // 5
+    // 5
+    // 5
+}
+
+void variablesOnHeapNoMove()
+{
+    std::cout << std::endl << std::endl;
+    std::cout << "Test on the heap with no move" << std::endl;
+
+    auto data = std::make_shared<int>(99);
+
+    auto fmo_ptr = std::make_unique<FunctionMemberObject>(print_data_func, data);
+    auto dmo_ptr = std::make_unique<DataMemberObject>(print_data_func, data);
+    auto nmo_ptr = std::make_unique<NoMemberObject>(print_data_func, data);
+
+    fmo_ptr->execute();
+    dmo_ptr->execute();
+    nmo_ptr->execute();
+
+    *data = 8;
+
+    fmo_ptr->execute();
+    dmo_ptr->execute();
+    nmo_ptr->execute();
+
+    // Output
+    // 99
+    // 99
+    // 99
+    // 8
+    // 8
+    // 8
+}
+
+void variablesOnStackWithMove()
+{
+    std::cout << std::endl << std::endl;
+    std::cout << "Test on the stack with moving variables" << std::endl;
+
+    auto data = std::make_shared<int>(99);
+
+    FunctionMemberObject fmo_test(print_data_func, data);
+    DataMemberObject dmo_test(print_data_func, data);
+    NoMemberObject nmo_test(print_data_func, data);
+
+    fmo_test.execute();
+    dmo_test.execute();
+    nmo_test.execute();
+
+    *data = -14;
+
+    FunctionMemberObject fmo_test_move = std::move(fmo_test);
+    DataMemberObject dmo_test_move     = std::move(dmo_test);
+    NoMemberObject nmo_test_move       = std::move(nmo_test);
+
+    fmo_test_move.execute();
+    dmo_test_move.execute();
+    nmo_test_move.execute();
+
+    *data = -15;
+
+    fmo_test_move.execute();
+    dmo_test_move.execute();
+    nmo_test_move.execute();
+
+    // Output
+    // 99
+    // 99
+    // 99
+    // -14
+    // -14
+    // -14
+    // -15
+    // -15
+    // -15
+}
+
+void variablesOnStackWithMoveVector()
+{
+    std::cout << std::endl << std::endl;
+    std::cout << "Test on the stack with moving vectors" << std::endl;
+
+    auto data = std::make_shared<int>(99);
+
+    std::vector<FunctionMemberObject> fmo_vector;
+    fmo_vector.push_back(std::move(FunctionMemberObject(print_data_func, data)));
+    std::vector<DataMemberObject> dmo_vector;
+    dmo_vector.push_back(std::move(DataMemberObject(print_data_func, data)));
+    std::vector<NoMemberObject> nmo_vector;
+    nmo_vector.push_back(std::move(NoMemberObject(print_data_func, data)));
+
+    //    fmo_vector.at(0).execute();
+    //    dmo_vector.at(0).execute();
+    nmo_vector.at(0).execute();
+
+    *data = -3;
+
+    //    fmo_vector.at(0).execute();
+    //    dmo_vector.at(0).execute();
+    nmo_vector.at(0).execute();
+
+    // Output
+    // std::bad_function_call / stops all printouts
+    // 0 (seems to be default-constructed data)
+    // 99
+    // std::bad_function_call / stops all printouts
+    // 0 (seems to be default-constructed data)
+    // -3
+
+    // Why does this not work for the DataMemberObject when it works
```

Comment:
If anyone has any ideas as to why vectors / containers behave differently than single-variables, I'm all ears.

Reply:
`std::vector` (and presumably any other resizable container) is by default heap-allocated, even if you declare the variable in a way that would normally allocate it on the stack.

Reply:
Oh right because it's resizable. I would totally believe moving between stack and heap would break the coroutine (and corrupt the program like I'm seeing).

Reply:
Will see if I can test this in a more explicit way.

Reply:
Note to self: try this
```
Coroutine foo();
auto foo_heap = std::make_unique<Coroutine>(); // empty
*foo_heap = std::move(foo);
// Or go the other way
```

Reply:
It was to do with heap allocation for resizable containers. Made better tests to make this clear.